### PR TITLE
Fixed 'Bug 60152 - Extension methods have wrong icon in CodeCompletion'

### PR DESCRIPTION
**Pull request by mkrueger**

### CODE REVIEW

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs [hunk 1]
```diff
@@ -90,10 +90,15 @@ public abstract class RoslynCompletionData : CompletionData
 				}
 				var modifier = GetItemModifier ();
 				var type = GetItemType ();
-				return "md-" + modifier + type;
+				var hash = modifier.GetHashCode() ^ type.GetHashCode();
+				if (!IconIdCache.ContainsKey(hash))
+					IconIdCache [hash] = "md-" + modifier + type;
```
Comment (slluis) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: This is not right, there is no guarantee `hash` is unique for a modifier and type.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs [hunk 2]
```diff
@@ -90,10 +90,15 @@ public abstract class RoslynCompletionData : CompletionData
 				}
 				var modifier = GetItemModifier ();
 				var type = GetItemType ();
-				return "md-" + modifier + type;
+				var hash = modifier.GetHashCode() ^ type.GetHashCode();
+				if (!IconIdCache.ContainsKey(hash))
```
Comment (Therzok) [inline main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: Are we sure all the possible values do not collide? Better have a unit test asserting this.

Comment (slluis) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: Is this optimization really worth it?

Comment (Therzok) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: We create this string for each icon in the completion widget, so if you have a 1 million RoslynCompletionData elements, you end up creating 1 million IconIds.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs (hunk 1 above)
Comment (mkrueger) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: It is - there are just a few strings I checked that the combos are unique.

File: main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs (hunk 2 above)
Comment (mkrueger) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: In general we should never do string allocations like that if they can be cached.

Comment (mkrueger) [reply main/src/core/MonoDevelop.Ide/MonoDevelop.Ide.CodeCompletion/RoslynCompletionData.cs]: @Therzok : Y I checked that there is no collision. Hashes can collide but we're speaking of 64 string combinations here and with such a low number a hash collision is very, very, very, very (insert some verys here) unlikely.

I added a unit test to ensure that…

**Pull request merged**